File: project_euler_unified/problem_306/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // Paper-strip game: count winning positions (G(n) != 0) for 1 <= n <= 10^6
    // Grundy values become periodic with period 34 starting at n=53

    const int LIMIT = 1000000;
    const int COMPUTE_UP_TO = 200;

    vector<int> G(COMPUTE_UP_TO + 1, 0);

    for (int n = 2; n <= COMPUTE_UP_TO; n++) {
        set<int> reachable;
        for (int i = 0; i <= n - 2; i++) {
            reachable.insert(G[i] ^ G[n - 2 - i]);
        }
        int mex = 0;
        while (reachable.count(mex)) mex++;
        G[n] = mex;
    }

    int period = 34;
    int offset = 53;

    // Verify periodicity
    for (int i = offset; i + period <= COMPUTE_UP_TO; i++) {
        assert(G[i] == G[i + period]);
    }

    // Count winning positions (G(n) != 0) for n = 1 to offset-1
    long long count = 0;
    for (int i = 1; i < offset && i <= LIMIT; i++) {
        if (G[i] != 0) count++;
    }

    if (LIMIT >= offset) {
        // Wins per period
        int wins_per_period = 0;
        for (int i = offset; i < offset + period; i++) {
            if (G[i] != 0) wins_per_period++;
        }

        long long remaining = (long long)(LIMIT) - offset + 1;
        long long full_periods = remaining / period;
        long long leftover = remaining % period;

        count += full_periods * wins_per_period;

        for (int i = 0; i < leftover; i++) {
            if (G[offset + i] != 0) count++;
        }
    }

    cout << count << endl;
    return 0;
}

File: project_euler_unified/problem_306/solution.py
"""
Problem 306: Paper-strip Game

Two players alternate turns on a strip of n white squares. Each turn, a player
picks two contiguous white squares and paints them black. The first player
unable to move loses.

For 1 <= n <= 1,000,000, how many values of n allow the first player to force a win?

Solution: Use Sprague-Grundy theory. When two adjacent squares are painted at
position i on a strip of n, the strip splits into independent subgames of sizes
i and n-2-i. Grundy values become periodic with period 34 starting at n=53.

Answer: 852938
"""

def solve():
    LIMIT = 1_000_000
    COMPUTE = 200

    G = [0] * (COMPUTE + 1)
    for n in range(2, COMPUTE + 1):
        reachable = set()
        for i in range(n - 1):
            reachable.add(G[i] ^ G[n - 2 - i])
        mex = 0
        while mex in reachable:
            mex += 1
        G[n] = mex

    period = 34
    offset = 53

    # Verify
    for i in range(offset, COMPUTE + 1 - period):
        assert G[i] == G[i + period]

    # Count winning positions for n = 1 to offset-1
    count = sum(1 for i in range(1, offset) if G[i] != 0)

    # Wins per period
    wins_per_period = sum(1 for i in range(offset, offset + period) if G[i] != 0)

    remaining = LIMIT - offset + 1
    full = remaining // period
    leftover = remaining % period

    count += full * wins_per_period
    count += sum(1 for i in range(leftover) if G[offset + i] != 0)

    print(count)
